# Segmenting and Clustering Neighborhoods in Toronto

## 1. Neighborhoods in Toronto: Dataframe

In this section, we extract data from  *https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M*, concerning administrative divisions in Toronto, and build a dataframe. 

We start by copy-paste the data into a csv file, but where different attributes are separated by a tab (rather then by a comma) and we pass the data into a pandas dataframe.

In [6]:
import pandas as pd

In [7]:
# tn - dataframe for Toronto's neighborhood data

tn=pd.read_csv('toronto.csv',delimiter='\t')

In [8]:
print(tn.shape,'\n')
tn.head(10)

(180, 3) 



Postal Code           Borough                                Neighbourhood
0         M1A      Not assigned                                 Not assigned
1         M2A      Not assigned                                 Not assigned
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                 Not assigned
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge

In [9]:
tn.describe()

Postal Code       Borough Neighbourhood
count          180           180           180
unique         180            11           101
top            M1M  Not assigned  Not assigned
freq             1            77            76

We see that there are no repeated postal codes in our dataframe, and that several postal codes have more than one neighborhood.

In [10]:
tn['Borough']=='Not assigned'

0       True
1       True
2      False
3      False
4      False
       ...  
175     True
176     True
177     True
178    False
179     True
Name: Borough, Length: 180, dtype: bool

We create a new dataframe where Postal Codes without an assigned Borough are cleared out.

In [11]:
# tnp - dataframe Toronto's neighborhood prime

tnp=tn[tn['Borough']!='Not assigned']
tnp.shape

(103, 3)

In [12]:
tnp.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [13]:
tnp.reset_index(drop=True,inplace=True)
tnp.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [14]:
tnp.describe()

Postal Code     Borough Neighbourhood
count          103         103           103
unique         103          10            99
top            M4G  North York     Downsview
freq             1          24             4

We check if there there are Boroughs without and assigned Neighborhood: 

In [15]:
tnpp=tnp[tnp['Neighbourhood']=='Not assigned']
tnpp.shape

(0, 3)

There are none!

Finally, we rename the third column as:

In [16]:
tnp.rename(columns={'Neighbourhood':'Neighborhood'},inplace=True)

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [17]:
tnp.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [168]:
tnp.shape

(103, 3)

The Toronto's Neighborhoods dataframe has 103 rows and 3 columns!

____________
## 2. Neighborhoods in Toronto: Coordinates

In this section, we attribute geographical coordinates for each Postal Code in Toronto.

We start by trying using geocoder:

In [18]:
!pip install geocoder

     |████████████████████████████████| 102kB 8.5MB/s ta 0:00:011


In [19]:
import geocoder

In [20]:

coords = None
postal_code='M5G'

#while(coord is None):
g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
coord=g.latlng

print(coord)

None


Geocoder does not work, therefore we use the results given by the dataset 'Geospatial_Coordinates.csv'.

In [22]:
df_coord=pd.read_csv('Geospatial_Coordinates.csv')
df_coord.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We make an inner join in python through the column keys 'Postal Code', in *tn* dataset, and 'Postal Code' in *df_coord* dataset.

In [23]:
# tnc - dataframe for Toronto's neighborhoods with the geographic coordinates

tnc = pd.merge(left=tnp, right=df_coord, left_on='Postal Code', right_on='Postal Code')
tnc.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Making a query:

In [24]:
tnc[tnc['Postal Code']=='M5G']

Postal Code           Borough        Neighborhood   Latitude  Longitude
24         M5G  Downtown Toronto  Central Bay Street  43.657952 -79.387383

In [25]:
tnc.shape

(103, 5)

The dataframe *tnc* has the same dimensions as *tnp*, meaning that it has been attributed a pair of coordinates to each Postal Code. 

In [26]:
tnc.to_csv('torontoc.csv')

___________________
## 3. Neighborhoods in Toronto: Clustering

In this section, we will collect data concerning different venues in Toronto city, and divide the city into clusters according to the frequency of those venues. Therefore, within each cluster in the city, the overall environment, with respect to venues, is more similar than outside the cluster.

We start my drawing a folium map showing the organization of the city in Boroughs.

In [27]:
!pip install folium

In [28]:
import folium 

In [29]:
!pip install geopy

     |████████████████████████████████| 112kB 6.4MB/s eta 0:00:01


In [30]:
from geopy.geocoders import Nominatim

Getting the geographical coordinates of Toronto city:

In [31]:
address = 'Toronto, CN'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6425637, -79.38708718320467.


In the map below, information concerning the Borough names, Neighborhoods names, and Postal Code are shown by clicking on the circles.

In [33]:
toronto_map=folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood, postalcode in zip(tnc['Latitude'], tnc['Longitude'],\
                                           tnc['Borough'], tnc['Neighborhood'],tnc['Postal Code']):
    label = '{}, {},{}'.format(neighborhood, borough,postalcode)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
#        color='blue',
        color='red',
        fill=True,
#        fill_color='#3186cc',
        fill_color='green',
        fill_opacity=0.7,
        parse_html=False).add_to(toronto_map)  
    
toronto_map

Now, let us explore the city of Toronto. We start by accessing a Foursquare account.

Let us imagine we live in the borough Downtown Toronto, so we wish to make a study of the nearby venues in our area.

In [35]:
tnc.loc[24,'Borough']

'Downtown Toronto'

We get the coordinates of Downtown Toronto:

In [36]:
myn_latitude=tnc.loc[24,'Latitude'] 
myn_longitude=tnc.loc[24,'Longitude']

myn_name = tnc.loc[24, 'Borough'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(myn_name, myn_latitude, myn_longitude))

Latitude and longitude values of Downtown Toronto are 43.6579524, -79.3873826.


Now, we make a Foursquare regular call to access the top 100 venues that are in Downtown Toronto, within a radius of 500 meters.

In [39]:
radius = 500
LIMIT=100 

url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, VERSION,myn_latitude,myn_longitude,radius, LIMIT)

In [37]:
import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [40]:
results=requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fec702d3577fa053cb86f31'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Bay Street Corridor',
  'headerFullLocation': 'Bay Street Corridor, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 62,
  'suggestedBounds': {'ne': {'lat': 43.6624524045, 'lng': -79.38117421839567},
   'sw': {'lat': 43.6534523955, 'lng': -79.39359098160432}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '537d4d6d498ec171ba22e7fe',
       'name': "Jimmy's Coffee",
       'location': {'address': '82 Gerrard Street W',
        'crossStreet': 'Gerrard & LaPlante',
        'lat': 43.65842123574496,
        'lng': -79.38561319551111,
        'label

Then we organize the results of our query in categories and build a dataframe with the nearby venues.

In [41]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [42]:

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name                 categories        lat        lng
0          Jimmy's Coffee                Coffee Shop  43.658421 -79.385613
1             Tim Hortons                Coffee Shop  43.658570 -79.385123
2  Somethin' 2 Talk About  Middle Eastern Restaurant  43.658395 -79.385338
3           Hailed Coffee                Coffee Shop  43.658833 -79.383684
4          NEO COFFEE BAR                Coffee Shop  43.660130 -79.385830

In [43]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

62 venues were returned by Foursquare.


In order to divide the city in clusters by venues, and therefore to find similar environments to ours, we now make Foursquare calls (1 call per Postal Code) to find the most frequent venues of all Boroughs in Toronto.  

In [44]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Borough', 
                  'Borough Latitude', 
                  'Borough Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Calling the function and building the dataframe *toronto_venues*.

In [45]:
# type your answer here
name=tnc['Borough']
lat=tnc['Latitude']
long=tnc['Longitude']
radius=500

toronto_venues=getNearbyVenues(name, lat, long, radius)
#manhattan_venues=pd.DataFrame(nearby_venues)

North York
North York
Downtown Toronto
North York
Downtown Toronto
Etobicoke
Scarborough
North York
East York
Downtown Toronto
North York
Etobicoke
Scarborough
North York
East York
Downtown Toronto
York
Etobicoke
Scarborough
East Toronto
Downtown Toronto
York
Scarborough
East York
Downtown Toronto
Downtown Toronto
Scarborough
North York
North York
East York
Downtown Toronto
West Toronto
Scarborough
North York
North York
East York
Downtown Toronto
West Toronto
Scarborough
North York
North York
East Toronto
Downtown Toronto
West Toronto
Scarborough
North York
North York
East Toronto
Downtown Toronto
North York
North York
Scarborough
North York
North York
East Toronto
North York
York
North York
Scarborough
North York
North York
Central Toronto
Central Toronto
York
York
Scarborough
North York
Central Toronto
Central Toronto
West Toronto
Etobicoke
Scarborough
North York
Central Toronto
Central Toronto
West Toronto
Mississauga
Etobicoke
Scarborough
Central Toronto
Downtown Toronto
West Toron

In [46]:
print('Number of venues in Toronto:', toronto_venues.shape[0],'\n')
toronto_venues.head()

Number of venues in Toronto: 2128 



Borough  Borough Latitude  Borough Longitude                   Venue  \
0  North York         43.753259         -79.329656         Brookbanks Park   
1  North York         43.753259         -79.329656           Variety Store   
2  North York         43.725882         -79.315572  Victoria Village Arena   
3  North York         43.725882         -79.315572               Portugril   
4  North York         43.725882         -79.315572             Tim Hortons   

   Venue Latitude  Venue Longitude         Venue Category  
0       43.751976       -79.332140                   Park  
1       43.751974       -79.333114      Food & Drink Shop  
2       43.723481       -79.315635           Hockey Arena  
3       43.725819       -79.312785  Portuguese Restaurant  
4       43.725517       -79.313103            Coffee Shop

Number of venues per Borough:

In [47]:
toronto_venues.drop(['Borough Latitude','Borough Longitude','Venue Latitude','Venue Longitude','Venue Category'],axis=1,inplace=False).groupby('Borough').count()

Venue
Borough                
Central Toronto     113
Downtown Toronto   1229
East Toronto        120
East York            73
Etobicoke            72
Mississauga          14
North York          245
Scarborough          92
West Toronto        151
York                 19

Number of different venue categories:

In [48]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 270 unique categories.


Building a dataframe, *toronto_onehot*, that organizes each venue by Borough and by category.

In [49]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Borough'] = toronto_venues['Borough'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Borough  Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0  North York                  0               0        0                   0   
1  North York                  0               0        0                   0   
2  North York                  0               0        0                   0   
3  North York                  0               0        0                   0   
4  North York                  0               0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  ...  Turkish Restaurant  \
0                    0  ...                   0   
1                    0  ...                   0   
2                    0  ...                   0   
3                    0  ...                   0   
4                    0  ...                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                0         0          0            0              0   
1                0         0          0            0              0   
2                0         0          0            0              0   
3                0         0          0            0              0   
4                0         0          0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 271 columns]

In [50]:
toronto_onehot.shape

(2128, 271)

Grouping the venues by Borough, and calculating the average of each venue category per Borough. This is the data that will be directly used for clustering.

In [51]:
toronto_grouped = toronto_onehot.groupby('Borough').mean().reset_index()
toronto_grouped

Borough  Accessories Store  Adult Boutique   Airport  \
0   Central Toronto           0.000000        0.000000  0.000000   
1  Downtown Toronto           0.000000        0.000814  0.000814   
2      East Toronto           0.000000        0.000000  0.000000   
3         East York           0.000000        0.000000  0.000000   
4         Etobicoke           0.000000        0.000000  0.000000   
5       Mississauga           0.000000        0.000000  0.000000   
6        North York           0.008163        0.000000  0.004082   
7       Scarborough           0.000000        0.000000  0.000000   
8      West Toronto           0.000000        0.000000  0.000000   
9              York           0.000000        0.000000  0.000000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0            0.000000      0.000000        0.000000         0.000000   
1            0.000814      0.000814        0.001627         0.001627   
2            0.000000      0.000000        0.000000         0.000000   
3            0.000000      0.000000        0.000000         0.000000   
4            0.000000      0.000000        0.000000         0.000000   
5            0.000000      0.000000        0.000000         0.000000   
6            0.000000      0.000000        0.000000         0.000000   
7            0.000000      0.000000        0.000000         0.000000   
8            0.000000      0.000000        0.000000         0.000000   
9            0.000000      0.000000        0.000000         0.000000   

   Airport Terminal  American Restaurant  ...  Turkish Restaurant  \
0          0.000000             0.017699  ...            0.000000   
1          0.001627             0.013832  ...            0.000000   
2          0.000000             0.025000  ...            0.000000   
3          0.000000             0.000000  ...            0.000000   
4          0.000000             0.013889  ...            0.000000   
5          0.000000             0.071429  ...            0.000000   
6          0.000000             0.008163  ...            0.000000   
7          0.000000             0.010870  ...            0.000000   
8          0.000000             0.000000  ...            0.000000   
9          0.000000             0.000000  ...            0.052632   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                       0.000000          0.000000               0.008850   
1                       0.011391          0.001627               0.004068   
2                       0.000000          0.000000               0.000000   
3                       0.000000          0.000000               0.000000   
4                       0.000000          0.000000               0.000000   
5                       0.000000          0.000000               0.000000   
6                       0.000000          0.004082               0.008163   
7                       0.000000          0.000000               0.010870   
8                       0.019868          0.000000               0.013245   
9                       0.000000          0.000000               0.000000   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0         0.000000  0.000000   0.000000     0.000000       0.000000   
1         0.000000  0.006509   0.000000     0.000000       0.000000   
2         0.000000  0.000000   0.000000     0.000000       0.000000   
3         0.013699  0.000000   0.000000     0.000000       0.000000   
4         0.000000  0.000000   0.000000     0.013889       0.000000   
5         0.000000  0.000000   0.000000     0.000000       0.000000   
6         0.000000  0.000000   0.000000     0.000000       0.008163   
7         0.000000  0.000000   0.000000     0.000000       0.000000   
8         0.000000  0.006623   0.006623     0.000000       0.000000   
9         0.000000  0.000000   0.000000     0.000000       0.052632   

   Yoga Studio  
0     0.008850  
1     0.005696  
2     0.016667  
3     0.0136

In [52]:
toronto_grouped.shape

(10, 271)

Listing the top 5 venues per Borough:

In [53]:
num_top_venues = 5

for hood in toronto_grouped['Borough']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Borough'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Central Toronto----
            venue  freq
0     Coffee Shop  0.09
1  Sandwich Place  0.06
2     Pizza Place  0.05
3            Café  0.05
4            Park  0.05


----Downtown Toronto----
                 venue  freq
0          Coffee Shop  0.10
1                 Café  0.05
2  Japanese Restaurant  0.03
3                Hotel  0.03
4           Restaurant  0.03


----East Toronto----
                venue  freq
0    Greek Restaurant  0.07
1         Coffee Shop  0.06
2  Italian Restaurant  0.04
3             Brewery  0.04
4      Ice Cream Shop  0.03


----East York----
            venue  freq
0            Bank  0.05
1     Coffee Shop  0.05
2     Pizza Place  0.04
3    Burger Joint  0.04
4  Sandwich Place  0.04


----Etobicoke----
                  venue  freq
0           Pizza Place  0.10
1        Sandwich Place  0.07
2              Pharmacy  0.06
3           Coffee Shop  0.06
4  Fast Food Restaurant  0.04


----Mississauga----
                      venue  freq
0               Coff

Building a dataframe with the the top 10 most common venues in each Borough:

In [54]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [55]:
import numpy as np

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Borough']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
borough_venues_sorted = pd.DataFrame(columns=columns)
borough_venues_sorted['Borough'] = toronto_grouped['Borough']

for ind in np.arange(toronto_grouped.shape[0]):
    borough_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

borough_venues_sorted.head()

Borough 1st Most Common Venue 2nd Most Common Venue  \
0   Central Toronto           Coffee Shop        Sandwich Place   
1  Downtown Toronto           Coffee Shop                  Café   
2      East Toronto      Greek Restaurant           Coffee Shop   
3         East York                  Bank           Coffee Shop   
4         Etobicoke           Pizza Place        Sandwich Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Pizza Place                  Park                  Café   
1            Restaurant                 Hotel   Japanese Restaurant   
2               Brewery    Italian Restaurant                  Park   
3          Burger Joint   Sporting Goods Shop        Sandwich Place   
4           Coffee Shop              Pharmacy  Fast Food Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Sushi Restaurant                   Gym            Restaurant   
1    Italian Restaurant                Bakery                  Park   
2        Ice Cream Shop                Bakery  Fast Food Restaurant   
3                  Park          Intersection           Pizza Place   
4                   Gym         Grocery Store   Fried Chicken Joint   

  9th Most Common Venue 10th Most Common Venue  
0        Clothing Store           Dessert Shop  
1    Seafood Restaurant               Beer Bar  
2                   Pub              Bookstore  
3            Restaurant              Pet Store  
4                Bakery         Discount Store

In [56]:
!pip install sklearn

  Stored in directory: /home/jupyterlab/.cache/pip/wheels/76/03/bb/589d421d27431bcd2c6da284d5f2286c8e3b2ea3cf1594c074
Successfully built sklearn


In [57]:
from sklearn.cluster import KMeans

Clustering the data using *toronto_grouped* dataframe. We choose seven clusters within the K-Means algorithm, i.e. k=7.

In [58]:
# set number of clusters
kclusters = 7

toronto_grouped_clustering = toronto_grouped.drop('Borough', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 4, 6, 5, 1, 0, 3, 4, 2], dtype=int32)

Attributing a Cluster label to each Postal Code, an integer running from 0 to 6. 

In [59]:
# add clustering labels
borough_venues_sorted.insert(0,'Cluster Labels', kmeans.labels_)

toronto_merged = tnc

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(borough_venues_sorted.set_index('Borough'), on='Borough')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656               0           Coffee Shop   
1  43.725882 -79.315572               0           Coffee Shop   
2  43.654260 -79.360636               0           Coffee Shop   
3  43.718518 -79.464763               0           Coffee Shop   
4  43.662301 -79.389494               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Clothing Store            Restaurant                  Park   
1        Clothing Store            Restaurant                  Park   
2                  Café            Restaurant                 Hotel   
3        Clothing Store            Restaurant                  Park   
4                  Café            Restaurant                 Hotel   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Japanese Restaurant        Sandwich Place                  Bank   
1   Japanese Restaurant        Sandwich Place                  Bank   
2   Japanese Restaurant    Italian Restaurant                Bakery   
3   Japanese Restaurant        Sandwich Place                  Bank   
4   Japanese Restaurant    Italian Restaurant                Bakery   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Pizza Place         Grocery Store          Shopping Mall  
1           Pizza Place         Grocery Store          Shopping Mall  
2                  Park    Seafood Restaurant               Beer Bar  
3           Pizza Place         Grocery Store          Shopping Mall  
4                  Park    Seafood Restaurant               Beer Bar

Finally, we display a map with the clustering results!

In [60]:
import matplotlib.cm as cm
import matplotlib.colors as colors
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [61]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'],toronto_merged['Longitude'],toronto_merged['Neighborhood'],toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
#        color='red',
        fill=True,
        fill_color=rainbow[cluster-1],
#        fill_color='green',
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

We can furthermore take a look on the distinguishable features of each cluster, by inspecting the top 10 venues of each cluster: 

In [62]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York               0           Coffee Shop   
1         North York               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3         North York               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
7         North York               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10        North York               0           Coffee Shop   
13        North York               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
20  Downtown Toronto               0           Coffee Shop   
24  Downtown Toronto               0           Coffee Shop   
25  Downtown Toronto               0           Coffee Shop   
27        North York               0           Coffee Shop   
28        North York               0           Coffee Shop   
30  Downtown Toronto               0           Coffee Shop   
33        North York               0           Coffee Shop   
34        North York               0           Coffee Shop   
36  Downtown Toronto               0           Coffee Shop   
39        North York               0           Coffee Shop   
40        North York               0           Coffee Shop   
42  Downtown Toronto               0           Coffee Shop   
45        North York               0           Coffee Shop   
46        North York               0           Coffee Shop   
48  Downtown Toronto               0           Coffee Shop   
49        North York               0           Coffee Shop   
50        North York               0           Coffee Shop   
52        North York               0           Coffee Shop   
53        North York               0           Coffee Shop   
55        North York               0           Coffee Shop   
57        North York               0           Coffee Shop   
59        North York               0           Coffee Shop   
60        North York               0           Coffee Shop   
61   Central Toronto               0           Coffee Shop   
62   Central Toronto               0           Coffee Shop   
66        North York               0           Coffee Shop   
67   Central Toronto               0           Coffee Shop   
68   Central Toronto               0           Coffee Shop   
72        North York               0           Coffee Shop   
73   Central Toronto               0           Coffee Shop   
74   Central Toronto               0           Coffee Shop   
79   Central Toronto               0           Coffee Shop   
80  Downtown Toronto               0           Coffee Shop   
83   Central Toronto               0           Coffee Shop   
84  Downtown Toronto               0           Coffee Shop   
86   Central Toronto               0           Coffee Shop   
87  Downtown Toronto               0           Coffee Shop   
91  Downtown Toronto               0           Coffee Shop   
92  Downtown Toronto               0           Coffee Shop   
96  Downtown Toronto               0           Coffee Shop   
97  Downtown Toronto               0           Coffee Shop   
99  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Clothing Store            Restaurant                  Park   
1         Clothing Store            Restaurant                  Park   
2                   Café            Restaurant                 Hotel   
3         Clothing Store            Restaurant                  Park   
4                   Café            Restaurant                 Hotel   
7         Clothing Store            Restaurant                  Park   
9                   Café            Restaurant                 Hotel   
10        Clothing Store            Restaurant                  Park   
13        Clothing Store            Restaurant                  Park   
15    

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
76  Mississauga               1           Coffee Shop                 Hotel   

   3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
76          Intersection                   Gym  Middle Eastern Restaurant   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
76  Mediterranean Restaurant   American Restaurant        Sandwich Place   

   9th Most Common Venue 10th Most Common Venue  
76   Fried Chicken Joint          Burrito Place

In [64]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
16    York               2                  Park           Coffee Shop   
21    York               2                  Park           Coffee Shop   
56    York               2                  Park           Coffee Shop   
63    York               2                  Park           Coffee Shop   
64    York               2                  Park           Coffee Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
16              Bus Line        Discount Store          Skating Rink   
21              Bus Line        Discount Store          Skating Rink   
56              Bus Line        Discount Store          Skating Rink   
63              Bus Line        Discount Store          Skating Rink   
64              Bus Line        Discount Store          Skating Rink   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
16  Caribbean Restaurant         Jewelry Store                  Pool   
21  Caribbean Restaurant         Jewelry Store                  Pool   
56  Caribbean Restaurant         Jewelry Store                  Pool   
63  Caribbean Restaurant         Jewelry Store                  Pool   
64  Caribbean Restaurant         Jewelry Store                  Pool   

   9th Most Common Venue 10th Most Common Venue  
16        Sandwich Place           Hockey Arena  
21        Sandwich Place           Hockey Arena  
56        Sandwich Place           Hockey Arena  
63        Sandwich Place           Hockey Arena  
64        Sandwich Place           Hockey Arena

In [65]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6   Scarborough               3          Intersection                  Bank   
12  Scarborough               3          Intersection                  Bank   
18  Scarborough               3          Intersection                  Bank   
22  Scarborough               3          Intersection                  Bank   
26  Scarborough               3          Intersection                  Bank   
32  Scarborough               3          Intersection                  Bank   
38  Scarborough               3          Intersection                  Bank   
44  Scarborough               3          Intersection                  Bank   
51  Scarborough               3          Intersection                  Bank   
58  Scarborough               3          Intersection                  Bank   
65  Scarborough               3          Intersection                  Bank   
71  Scarborough               3          Intersection                  Bank   
78  Scarborough               3          Intersection                  Bank   
82  Scarborough               3          Intersection                  Bank   
85  Scarborough               3          Intersection                  Bank   
90  Scarborough               3          Intersection                  Bank   
95  Scarborough               3          Intersection                  Bank   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6            Coffee Shop                Bakery  Fast Food Restaurant   
12           Coffee Shop                Bakery  Fast Food Restaurant   
18           Coffee Shop                Bakery  Fast Food Restaurant   
22           Coffee Shop                Bakery  Fast Food Restaurant   
26           Coffee Shop                Bakery  Fast Food Restaurant   
32           Coffee Shop                Bakery  Fast Food Restaurant   
38           Coffee Shop                Bakery  Fast Food Restaurant   
44           Coffee Shop                Bakery  Fast Food Restaurant   
51           Coffee Shop                Bakery  Fast Food Restaurant   
58           Coffee Shop                Bakery  Fast Food Restaurant   
65           Coffee Shop                Bakery  Fast Food Restaurant   
71           Coffee Shop                Bakery  Fast Food Restaurant   
78           Coffee Shop                Bakery  Fast Food Restaurant   
82           Coffee Shop                Bakery  Fast Food Restaurant   
85           Coffee Shop                Bakery  Fast Food Restaurant   
90           Coffee Shop                Bakery  Fast Food Restaurant   
95           Coffee Shop                Bakery  Fast Food Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6     Chinese Restaurant              Pharmacy           Pizza Place   
12    Chinese Restaurant              Pharmacy           Pizza Place   
18    Chinese Restaurant              Pharmacy           Pizza Place   
22    Chinese Restaurant              Pharmacy           Pizza Place   
26    Chinese Restaurant              Pharmacy           Pizza Place   
32    Chinese Restaurant              Pharmacy           Pizza Place   
38    Chinese Restaurant              Pharmacy           Pizza Place   
44    Chinese Restaurant              Pharmacy           Pizza Place   
51    Chinese Restaurant              Pharmacy           Pizza Place   
58    Chinese Restaurant              Pharmacy           Pizza Place   
65    Chinese Restaurant              Pharmacy           Pizza Place   
71    Chinese Restaurant              Pharmacy           Pizza Place   
78    Chinese Restaurant              Pharmacy           Pizza Place   
82    Chinese Restaurant              Pharmacy           Pizza Place   
85    Chinese Restaurant              Pharmacy           Pizza Place   
90    Chinese Restaurant              Pharmacy           Pizza Place   
95    Chinese Restaurant              Pharmacy           Pizza P

In [66]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19   East Toronto               4      Greek Restaurant           Coffee Shop   
31   West Toronto               4                  Café                   Bar   
37   West Toronto               4                  Café                   Bar   
41   East Toronto               4      Greek Restaurant           Coffee Shop   
43   West Toronto               4                  Café                   Bar   
47   East Toronto               4      Greek Restaurant           Coffee Shop   
54   East Toronto               4      Greek Restaurant           Coffee Shop   
69   West Toronto               4                  Café                   Bar   
75   West Toronto               4                  Café                   Bar   
81   West Toronto               4                  Café                   Bar   
100  East Toronto               4      Greek Restaurant           Coffee Shop   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19                Brewery    Italian Restaurant                  Park   
31            Coffee Shop    Italian Restaurant                Bakery   
37            Coffee Shop    Italian Restaurant                Bakery   
41                Brewery    Italian Restaurant                  Park   
43            Coffee Shop    Italian Restaurant                Bakery   
47                Brewery    Italian Restaurant                  Park   
54                Brewery    Italian Restaurant                  Park   
69            Coffee Shop    Italian Restaurant                Bakery   
75            Coffee Shop    Italian Restaurant                Bakery   
81            Coffee Shop    Italian Restaurant                Bakery   
100               Brewery    Italian Restaurant                  Park   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19         Ice Cream Shop                Bakery  Fast Food Restaurant   
31             Restaurant        Breakfast Spot                 Diner   
37             Restaurant        Breakfast Spot                 Diner   
41         Ice Cream Shop                Bakery  Fast Food Restaurant   
43             Restaurant        Breakfast Spot                 Diner   
47         Ice Cream Shop                Bakery  Fast Food Restaurant   
54         Ice Cream Shop                Bakery  Fast Food Restaurant   
69             Restaurant        Breakfast Spot                 Diner   
75             Restaurant        Breakfast Spot                 Diner   
81             Restaurant        Breakfast Spot                 Diner   
100        Ice Cream Shop                Bakery  Fast Food Restaurant   

    9th Most Common Venue 10th Most Common Venue  
19                    Pub              Bookstore  
31              Bookstore            Pizza Place  
37              Bookstore            Pizza Place  
41                    Pub              Bookstore  
43              Bookstore            Pizza Place  
47                    Pub              Bookstore  
54                    Pub              Bookstore  
69              Bookstore            Pizza Place  
75              Bookstore            Pizza Place  
81              Bookstore            Pizza Place  
100                   Pub              Bookstore

In [67]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5    Etobicoke               5           Pizza Place        Sandwich Place   
11   Etobicoke               5           Pizza Place        Sandwich Place   
17   Etobicoke               5           Pizza Place        Sandwich Place   
70   Etobicoke               5           Pizza Place        Sandwich Place   
77   Etobicoke               5           Pizza Place        Sandwich Place   
88   Etobicoke               5           Pizza Place        Sandwich Place   
89   Etobicoke               5           Pizza Place        Sandwich Place   
93   Etobicoke               5           Pizza Place        Sandwich Place   
94   Etobicoke               5           Pizza Place        Sandwich Place   
98   Etobicoke               5           Pizza Place        Sandwich Place   
101  Etobicoke               5           Pizza Place        Sandwich Place   
102  Etobicoke               5           Pizza Place        Sandwich Place   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5             Coffee Shop              Pharmacy  Fast Food Restaurant   
11            Coffee Shop              Pharmacy  Fast Food Restaurant   
17            Coffee Shop              Pharmacy  Fast Food Restaurant   
70            Coffee Shop              Pharmacy  Fast Food Restaurant   
77            Coffee Shop              Pharmacy  Fast Food Restaurant   
88            Coffee Shop              Pharmacy  Fast Food Restaurant   
89            Coffee Shop              Pharmacy  Fast Food Restaurant   
93            Coffee Shop              Pharmacy  Fast Food Restaurant   
94            Coffee Shop              Pharmacy  Fast Food Restaurant   
98            Coffee Shop              Pharmacy  Fast Food Restaurant   
101           Coffee Shop              Pharmacy  Fast Food Restaurant   
102           Coffee Shop              Pharmacy  Fast Food Restaurant   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5                     Gym         Grocery Store   Fried Chicken Joint   
11                    Gym         Grocery Store   Fried Chicken Joint   
17                    Gym         Grocery Store   Fried Chicken Joint   
70                    Gym         Grocery Store   Fried Chicken Joint   
77                    Gym         Grocery Store   Fried Chicken Joint   
88                    Gym         Grocery Store   Fried Chicken Joint   
89                    Gym         Grocery Store   Fried Chicken Joint   
93                    Gym         Grocery Store   Fried Chicken Joint   
94                    Gym         Grocery Store   Fried Chicken Joint   
98                    Gym         Grocery Store   Fried Chicken Joint   
101                   Gym         Grocery Store   Fried Chicken Joint   
102                   Gym         Grocery Store   Fried Chicken Joint   

    9th Most Common Venue 10th Most Common Venue  
5                  Bakery         Discount Store  
11                 Bakery         Discount Store  
17                 Bakery         Discount Store  
70                 Bakery         Discount Store  
77                 Bakery         Discount Store  
88                 Bakery         Discount Store  
89                 Bakery         Discount Store  
93                 Bakery         Discount Store  
94                 Bakery         Discount Store  
98                 Bakery         Discount Store  
101                Bakery         Discount Store  
102                Bakery         Discount Store

In [68]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 6, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
8   East York               6                  Bank           Coffee Shop   
14  East York               6                  Bank           Coffee Shop   
23  East York               6                  Bank           Coffee Shop   
29  East York               6                  Bank           Coffee Shop   
35  East York               6                  Bank           Coffee Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
8           Burger Joint   Sporting Goods Shop        Sandwich Place   
14          Burger Joint   Sporting Goods Shop        Sandwich Place   
23          Burger Joint   Sporting Goods Shop        Sandwich Place   
29          Burger Joint   Sporting Goods Shop        Sandwich Place   
35          Burger Joint   Sporting Goods Shop        Sandwich Place   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
8                   Park          Intersection           Pizza Place   
14                  Park          Intersection           Pizza Place   
23                  Park          Intersection           Pizza Place   
29                  Park          Intersection           Pizza Place   
35                  Park          Intersection           Pizza Place   

   9th Most Common Venue 10th Most Common Venue  
8             Restaurant              Pet Store  
14            Restaurant              Pet Store  
23            Restaurant              Pet Store  
29            Restaurant              Pet Store  
35            Restaurant              Pet Store

We can verify the similarities among cluster elements, in most cases, the top 10 venues are basically the same!

Moreover, we can also check that our Downtown Toronto Borough lies in Cluster 0, where Coffe and Restaurant is a priority, specially Japanese and Pizza. Since this Cluster covers a large area, it is easier to find similar environments within the city!

The end.